octobot_trading/producers/simulator/ohlcv_updater_simulator.py

```diff
@@ -44,22 +51,36 @@ async def wait_for_initialization(self, timeout=OHLCVUpdater.OHLCV_INITIALIZATIO
     async def handle_timestamp(self, timestamp, **kwargs):
         if not self.is_initialized:
             await self.wait_for_initialization()
-        if self.last_timestamp_pushed == 0:
-            self.last_timestamp_pushed = timestamp
 
         try:
             for time_frame in self.channel.exchange_manager.exchange_config.traded_time_frames:
                 for pair in self.channel.exchange_manager.exchange_config.traded_symbol_pairs:
-                    # use timestamp - 1 for superior timestamp to avoid select of a future candle
+                    # use last_timestamp_pushed + 1 for inferior timestamp to avoid select of an already selected candle
                     # (selection is <= and >=)
+                    # use timestamp + time_frame seconds to include the future candle that will be sorted in exchange
+                    # simulator for later uses.
                     ohlcv_data: list = await self.exchange_data_importer.get_ohlcv_from_timestamps(
                         exchange_name=self.exchange_name,
                         symbol=pair,
                         time_frame=time_frame,
-                        inferior_timestamp=self.last_timestamp_pushed,
-                        superior_timestamp=timestamp - 1)
+                        inferior_timestamp=self.last_timestamp_pushed + 1,
+                        superior_timestamp=timestamp + self.time_frames_to_second[time_frame])
                     if ohlcv_data:
-                        await self.push(time_frame, pair, [ohlcv[-1] for ohlcv in ohlcv_data], partial=True)
+                        current_candle_index = 0
```

Comment:
I don't really understand the role off `current_candle_index`

Reply:
It's used to know where the candle to push is and not push the future candle instead

Reply:
Ok, that was a bit confusing with `ohlcv_data[0]`. If I understand well `ohlcv_data[0]` could have been `ohlcv_data[current_candle_index]` ?

Reply:
When there is no future candle yes (when there are missing candles for 2h between 12 and 14h ex or at the very end of the datafile

Reply:
right :+1: